### ISSUE TITLE
JsonTypeInfo DEDUCTION causes erroneous WRAPPER_ARRAY with JsonValue

### ISSUE THREAD

**Issue by chrylis** [FasterXML/jackson-databind#3579]
I'm working on a JSON:API utility library and experimenting with ways to handle its remarkably flexible polymorphism. In this case, I am trying to model [links](https://jsonapi.org/format/1.1/#document-links), where the value can be either an `object` or a `string` representing a URI.

My initial attempt (which would not entirely work for other reasons) was as follows:

```
@JsonTypeInfo(use = DEDUCTION)
@JsonSubTypes({
    @Type(JsonApiLink.BareUri.class),
    @Type(JsonApiLink.LinkObject.class),
})
public abstract class JsonApiLink {
    public abstract URI getHref();

    @Data
    @EqualsAndHashCode(callSuper = false)
    public static class BareUri extends JsonApiLink {
        @JsonValue
        URI href;
    }


    @Data
    @EqualsAndHashCode(callSuper = false)
    public static class LinkObject extends JsonApiLink {
        URI href;

        String rel;
    }
}
```

Invoking `jackson.writeValueAsString(new LinkObject(href: uri))` produces the expected "JSON object in a Java string" with no type information, but when I attempt to call `jackson.writeValueAsString(new BareUri(uri))`, I get the following output:

```
["com.example.jsonapi.links.JsonApiLink$BareUri","https://links.test/d3a99621-971b-40b1-b8e4-cbbdd3299602"]
```

It appears that `TypeSerializerRerouter` is invoked from `ToStringSerializerBase#serializeWithType`, swaps the `@JsonValue`'s strategy for the subtype's strategy, and then `JsonGenerator` forgets the `DEDUCTION` instruction and imposes a default of `WRAPPER_ARRAY` (`JsonGenerator.java:1934` in 2.13.2).

**Comment by smilep** [comment]
Similar issue. Reproducible when using an interface type whose implementation is an enum.

**Comment by cowtowncoder** [comment]
I _think_ this is dup of #3711 (or the original one, but since I marked 3711 via fix, consider that as the main issue).

@chrylis If you have time, would appreciate verification for fix in `2.14` branch (and merged into `2.15`).
